lib/beaker-hostgenerator/generator.rb

```diff
@@ -32,43 +37,38 @@ def generate(tokens, options)
           next
         end
 
-        node_info = __parse_node_info_token(token)
+        node_info = parse_node_info_token(token)
+
+        # Build node host name
+        platform = "#{ostype}-#{node_info['bits']}"
+        host_name = "#{platform}-#{nodeid[ostype]}"
 
+        node_info['platform'] = platform
         node_info['ostype'] = ostype
         node_info['nodeid'] = nodeid[ostype]
 
-        host_config = base_host_config()
-
-        [:pe_dir, :pe_ver, :pe_upgrade_dir, :pe_upgrade_ver].each do |option|
-          if options[option]
-            host_config[option.to_s] = options[option]
-          end
-        end
+        host_config = base_host_config(options)
 
+        # Delegate to the hypervisor
         hypervisor = BeakerHostGenerator::Hypervisor.create(node_info, options)
-        host_name, host_config =
-                   hypervisor.generate_node(node_info, host_config, bhg_version)
+        host_config = hypervisor.generate_node(node_info, host_config, bhg_version)
         config['CONFIG'].deep_merge!(hypervisor.global_config())
 
+        # Merge in any arbitrary key-value host settings. Treat the 'hostname'
+        # setting specially, and don't merge it in as an arbitrary setting.
+        arbitrary_settings = node_info['host_settings']
+        if arbitrary_settings['hostname']
+          host_name = arbitrary_settings['hostname']
+          arbitrary_settings.delete('hostname')
+        end
+        host_config.merge!(arbitrary_settings)
```

Comment:
I can't tell clearly from this diff context, but is there any risk of direct changes to `host_config` impacting resources held by outside callers? (maybe whatever is being passed around by `base_host_config`)

If so, copying the hash early would get rid of any action-at-a-distance problems.

It's harder to reason about, but it might be safer to just copy the hash anyway, in case under maintenance a reuse which could be acted upon at a distance might get introduced at some point.